Panic runtime error: index out of range when running 6 node tpc-c without range merging

**Describe the problem**

Node died while running tpc-c on a six node cluster with range merging turned off. In the shell I saw 

```
Error: error in payment: EOF
Error:  exit status 1
```
**To Reproduce**

```
export CLUSTER=andy-base
roachprod create $CLUSTER -n 7 --clouds=aws  --aws-machine-type-ssd=c5d.4xlarge
roachprod run $CLUSTER:1-6 -- 'sudo umount /mnt/data1; sudo mount -o discard,defaults,nobarrier /dev/nvme1n1 /mnt/data1/; mount | grep /mnt/data1'
roachprod stage $CLUSTER:1-6 cockroach
roachprod stage $CLUSTER:7 workload
roachprod start $CLUSTER:1-6
roachprod sql $CLUSTER:1 -- -e 'SET CLUSTER SETTING kv.range_merge.queue_enabled = false'
roachprod adminurl --open $CLUSTER:1
roachprod run $CLUSTER:1 -- "./cockroach workload fixtures import tpcc --warehouses=5000 --db=tpcc"
roachprod run $CLUSTER:7 "./workload run tpcc --ramp=5m --warehouses=3400 --duration=15m --split --scatter {pgurl:1-6}"
```

**Expected behavior**
Completing tpc-c

**Additional data / screenshots**
![image](https://user-images.githubusercontent.com/22278911/51719336-15960280-2017-11e9-814e-87d25bc0c6d2.png)

**Environment:**
v2.2.0-alpha.20181217-820-g645c0c9

From log to stderr
https://gist.github.com/awoods187/a44b3505f7921aa152d8c0b488afdccc
 
```
ERROR: [n5,client=172.31.42.127:50898,user=root,txn=44987ebc] a panic has occurred!
runtime error: index out of range
```

I believe this crash here might be a repro:
https://github.com/cockroachdb/cockroach/issues/34695#issuecomment-469027213

```
panic: batch size 196 -> 194 bytes
re-marshaled protobuf:
00000000  0a 9d 01 0a 00 12 06 08  04 10 04 18 03 18 47 2a  |..............G*|
00000010  87 01 0a 22 0a 10 d0 96  0f dd 5a a8 41 65 b0 b2  |..."......Z.Ae..|
00000020  01 80 af fa 3c 66 2a 0a  08 8a e7 e3 dc 8c 8e 9e  |....<f*.........|
00000030  c4 15 30 87 b0 05 12 07  73 71 6c 20 74 78 6e 2a  |..0.....sql txn*|
00000040  0a 08 8a e7 e3 dc 8c 8e  9e c4 15 32 0a 08 8a e7  |...........2....|
00000050  e3 dc 8c 8e 9e c4 15 3a  0a 08 8a b1 99 cb 8e 8e  |.......:........|
00000060  9e c4 15 42 0e 08 01 12  0a 08 8a e7 e3 dc 8c 8e  |...B............|
00000070  9e c4 15 42 0e 08 04 12  0a 08 e5 ed d3 ad 8d 8e  |...B............|
00000080  9e c4 15 42 10 08 05 12  0c 08 ab ee f5 ab 8d 8e  |...B............|
00000090  9e c4 15 10 0a 72 00 7a  00 40 8f 4e 50 01 58 01  |.....r.z.@.NP.X.|
000000a0  12 20 3a 1e 0a 1a 1a 0b  bd 89 fd 05 fb 76 e8 54  |. :..........v.T|
000000b0  ce 80 06 22 0b bd 89 fd  05 fb 76 e8 55 63 00 01  |..."......v.Uc..|
000000c0  20 01                                             | .|

original panic:  <nil>
 [recovered]
	panic: batch size 196 -> 194 bytes
re-marshaled protobuf:
00000000  0a 9d 01 0a 00 12 06 08  04 10 04 18 03 18 47 2a  |..............G*|
00000010  87 01 0a 22 0a 10 d0 96  0f dd 5a a8 41 65 b0 b2  |..."......Z.Ae..|
00000020  01 80 af fa 3c 66 2a 0a  08 8a e7 e3 dc 8c 8e 9e  |....<f*.........|
00000030  c4 15 30 87 b0 05 12 07  73 71 6c 20 74 78 6e 2a  |..0.....sql txn*|
00000040  0a 08 8a e7 e3 dc 8c 8e  9e c4 15 32 0a 08 8a e7  |...........2....|
00000050  e3 dc 8c 8e 9e c4 15 3a  0a 08 8a b1 99 cb 8e 8e  |.......:........|
00000060  9e c4 15 42 0e 08 01 12  0a 08 8a e7 e3 dc 8c 8e  |...B............|
00000070  9e c4 15 42 0e 08 04 12  0a 08 e5 ed d3 ad 8d 8e  |...B............|
00000080  9e c4 15 42 10 08 05 12  0c 08 ab ee f5 ab 8d 8e  |...B............|
00000090  9e c4 15 10 0a 72 00 7a  00 40 8f 4e 50 01 58 01  |.....r.z.@.NP.X.|
000000a0  12 20 3a 1e 0a 1a 1a 0b  bd 89 fd 05 fb 76 e8 54  |. :..........v.T|
000000b0  ce 80 06 22 0b bd 89 fd  05 fb 76 e8 55 63 00 01  |..."......v.Uc..|
000000c0  20 01                                             | .|

original panic:  <nil>


goroutine 66204 [running]:
github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).divideAndSendBatchToRanges.func1(0xc009678318, 0xc009678608, 0xc009678548, 0xc009678600, 0xc00967827f, 0xc009678300, 0xc009678284)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:825 +0x50e
panic(0x2d635a0, 0xc00915b8c0)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/kv.withMarshalingDebugging.func1(0xc009502f80, 0xc4)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/transport.go:181 +0x264
github.com/cockroachdb/cockroach/pkg/kv.withMarshalingDebugging(0x3a16420, 0xc007b8b1d0, 0x0, 0x0, 0x400000004, 0x3, 0x47, 0x0, 0xc0088e4a00, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/transport.go:185 +0xac
github.com/cockroachdb/cockroach/pkg/kv.(*grpcTransport).SendNext(0xc007b8b1a0, 0x3a16420, 0xc00885bef0, 0x0, 0x0, 0x400000004, 0x3, 0x47, 0x0, 0xc0088e4a00, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/transport.go:202 +0x211
github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendToReplicas(0xc00020d7a0, 0x3a16420, 0xc00885bef0, 0xc00020d7f0, 0x47, 0xc00750a6e0, 0x3, 0x3, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1363 +0x2d3
github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendRPC(0xc00020d7a0, 0x3a16420, 0xc00885bef0, 0x47, 0xc00750a6e0, 0x3, 0x3, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:416 +0x244
github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendSingleRange(0xc00020d7a0, 0x3a16420, 0xc00885bef0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0088e4a00, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:496 +0x221
github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendPartialBatch(0xc00020d7a0, 0x3a16420, 0xc00885bef0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0088e4a00, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1139 +0x322
github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).divideAndSendBatchToRanges(0xc00020d7a0, 0x3a16420, 0xc00885bef0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0088e4a00, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:962 +0x8b3
github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).Send(0xc00020d7a0, 0x3a16420, 0xc00885bef0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00819da00, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:710 +0x48b
github.com/cockroachdb/cockroach/pkg/kv.(*txnLockGatekeeper).SendLocked(0xc009939f10, 0x3a16420, 0xc00885bef0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00819da00, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:233 +0xe8
github.com/cockroachdb/cockroach/pkg/kv.(*txnSpanRefresher).sendLockedWithRefreshAttempts(0xc009939e48, 0x3a16420, 0xc00885bef0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00819da00, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_interceptor_span_refresher.go:160 +0x83
github.com/cockroachdb/cockroach/pkg/kv.(*txnSpanRefresher).SendLocked(0xc009939e48, 0x3a16420, 0xc00885bef0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00819da00, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_interceptor_span_refresher.go:101 +0xf9
github.com/cockroachdb/cockroach/pkg/kv.(*txnPipeliner).SendLocked(0xc009939dc0, 0x3a16420, 0xc00885bef0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00819da00, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_interceptor_pipeliner.go:182 +0xf9
github.com/cockroachdb/cockroach/pkg/kv.(*txnSeqNumAllocator).SendLocked(0xc009939d68, 0x3a16420, 0xc00885bef0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00819da00, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_interceptor_seq_num_allocator.go:92 +0x23b
github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).Send(0xc009939b00, 0x3a16420, 0xc00885bef0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:698 +0x591
github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).sendUsingSender(0xc000443200, 0x3a16420, 0xc00885bec0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:622 +0x119
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Send(0xc0079034d0, 0x3a16420, 0xc00885bec0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:790 +0x13c
github.com/cockroachdb/cockroach/pkg/sql/row.(*txnKVFetcher).fetch(0xc007ef6d00, 0x3a16420, 0xc00885bec0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/kv_batch_fetcher.go:242 +0x626
github.com/cockroachdb/cockroach/pkg/sql/row.(*txnKVFetcher).nextBatch(0xc007ef6d00, 0x3a16420, 0xc00885bec0, 0xc007ca6e80, 0x30, 0xc007ca4ba0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/row/kv_batch_fetcher.go:326 +0x1dd
```

--- a/pkg/cmd/roachtest/kv.go
+++ b/pkg/cmd/roachtest/kv.go
@@ -100,8 +100,7 @@ func registerKV(r *registry) {
 		// Configs with large nodes.
 		{nodes: 3, cpus: 96, readPercent: 0},
 		{nodes: 3, cpus: 96, readPercent: 95},
-		// Skipped: https://github.com/cockroachdb/cockroach/issues/34241.
-		// {nodes: 4, cpus: 96, readPercent: 50, batchSize: 64},
+		{nodes: 4, cpus: 96, readPercent: 50, batchSize: 64},
 
 		// Configs with encryption.
 		{nodes: 1, cpus: 8, readPercent: 0, encryption: true},

--- a/pkg/internal/client/txn_test.go
+++ b/pkg/internal/client/txn_test.go
@@ -385,7 +385,6 @@ func TestSetPriority(t *testing.T) {
 				}
 
 				br := &roachpb.BatchResponse{}
-				br.Txn = &roachpb.Transaction{}
 				br.Txn.Update(ba.Txn) // copy
 				return br, nil
 			}), clock)

--- a/pkg/kv/dist_sender.go
+++ b/pkg/kv/dist_sender.go
@@ -967,7 +967,9 @@ func (ds *DistSender) divideAndSendBatchToRanges(
 			// Update the transaction from the response. Note that this wouldn't happen
 			// on the asynchronous path, but if we have newer information it's good to
 			// use it.
-			ba.UpdateTxn(resp.reply.Txn)
+			if !lastRange {
+				ba.UpdateTxn(resp.reply.Txn)
+			}
 
 			mightStopEarly := ba.MaxSpanRequestKeys > 0 || stopAtRangeBoundary
 			// Check whether we've received enough responses to exit query loop.

--- a/pkg/kv/transport.go
+++ b/pkg/kv/transport.go
@@ -17,10 +17,7 @@ package kv
 
 import (
 	"context"
-	"encoding/hex"
-	"fmt"
 	"sort"
-	"strings"
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/internal/client"
@@ -28,7 +25,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/rpc"
 	"github.com/cockroachdb/cockroach/pkg/rpc/nodedialer"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
-	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
 	"github.com/opentracing/opentracing-go"
@@ -156,38 +152,6 @@ func (gt *grpcTransport) maybeResurrectRetryablesLocked() bool {
 	return len(resurrect) > 0
 }
 
-func withMarshalingDebugging(ctx context.Context, ba roachpb.BatchRequest, f func()) {
-	nPre := ba.Size()
-	defer func() {
-		// TODO(ajwerner): re-enable the pre-emptive panic case below when the sizes
-		// do not match. The case is being disabled temporarily to reduce the
-		// rate of panics in the upcoming release. A more holistic fix which
-		// eliminates the shallow copies of transactions is coming soon.
-		nPost := ba.Size()
-		if r := recover(); r != nil /* || nPre != nPost */ {
-			var buf strings.Builder
-			_, _ = fmt.Fprintf(&buf, "batch size %d -> %d bytes\n", nPre, nPost)
-			func() {
-				defer func() {
-					if rInner := recover(); rInner != nil {
-						_, _ = fmt.Fprintln(&buf, "panic while re-marshaling:", rInner)
-					}
-				}()
-				data, mErr := protoutil.Marshal(&ba)
-				if mErr != nil {
-					_, _ = fmt.Fprintln(&buf, "while re-marshaling:", mErr)
-				} else {
-					_, _ = fmt.Fprintln(&buf, "re-marshaled protobuf:")
-					_, _ = fmt.Fprintln(&buf, hex.Dump(data))
-				}
-			}()
-			_, _ = fmt.Fprintln(&buf, "original panic: ", r)
-			panic(buf.String())
-		}
-	}()
-	f()
-}
-
 // SendNext invokes the specified RPC on the supplied client when the
 // client is ready. On success, the reply is sent on the channel;
 // otherwise an error is sent.
@@ -201,11 +165,7 @@ func (gt *grpcTransport) SendNext(
 	}
 
 	ba.Replica = client.replica
-	var reply *roachpb.BatchResponse
-
-	withMarshalingDebugging(ctx, ba, func() {
-		reply, err = gt.sendBatch(ctx, client.replica.NodeID, iface, ba)
-	})
+	reply, err := gt.sendBatch(ctx, client.replica.NodeID, iface, ba)
 
 	// NotLeaseHolderErrors can be retried.
 	var retryable bool

--- a/pkg/kv/transport_test.go
+++ b/pkg/kv/transport_test.go
@@ -26,7 +26,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
 	opentracing "github.com/opentracing/opentracing-go"
-	"github.com/stretchr/testify/assert"
 	"google.golang.org/grpc"
 )
 
@@ -172,31 +171,4 @@ func (m *mockInternalClient) RangeFeed(
 	ctx context.Context, in *roachpb.RangeFeedRequest, opts ...grpc.CallOption,
 ) (roachpb.Internal_RangeFeedClient, error) {
 	return nil, fmt.Errorf("unsupported RangeFeed call")
-}
-
-func TestWithMarshalingDebugging(t *testing.T) {
-	defer leaktest.AfterTest(t)()
-	t.Skip(fmt.Sprintf("Skipped until #34241 is resolved"))
-
-	ctx := context.Background()
-
-	exp := `batch size 23 -> 28 bytes
-re-marshaled protobuf:
-00000000  0a 0a 0a 00 12 06 08 00  10 00 18 00 12 0e 3a 0c  |..............:.|
-00000010  0a 0a 1a 03 66 6f 6f 22  03 62 61 72              |....foo".bar|
-
-original panic:  <nil>
-`
-
-	assert.PanicsWithValue(t, exp, func() {
-		var ba roachpb.BatchRequest
-		ba.Add(&roachpb.ScanRequest{
-			RequestHeader: roachpb.RequestHeader{
-				Key: []byte("foo"),
-			},
-		})
-		withMarshalingDebugging(ctx, ba, func() {
-			ba.Requests[0].GetInner().(*roachpb.ScanRequest).EndKey = roachpb.Key("bar")
-		})
-	})
 }

--- a/pkg/kv/txn_coord_sender.go
+++ b/pkg/kv/txn_coord_sender.go
@@ -905,8 +905,7 @@ func (tc *TxnCoordSender) UpdateStateOnRemoteRetryableErr(
 	// handleRetryableErrLocked().
 	if err.Transaction.ID == txnID {
 		// This is where we get a new epoch.
-		cp := err.Transaction.Clone()
-		tc.mu.txn.Update(&cp)
+		tc.mu.txn.Update(&err.Transaction)
 	}
 	return roachpb.NewError(err)
 }
@@ -1036,8 +1035,7 @@ func (tc *TxnCoordSender) updateStateLocked(
 			tc.mu.storedErr = pErr
 
 			// Cleanup.
-			cp := pErr.GetTxn().Clone()
-			tc.mu.txn.Update(&cp)
+			tc.mu.txn.Update(pErr.GetTxn())
 			tc.cleanupTxnLocked(ctx)
 			return pErr
 		}
@@ -1056,8 +1054,7 @@ func (tc *TxnCoordSender) updateStateLocked(
 		// handleRetryableErrLocked().
 		if err.Transaction.ID == ba.Txn.ID {
 			// This is where we get a new epoch.
-			cp := err.Transaction.Clone()
-			tc.mu.txn.Update(&cp)
+			tc.mu.txn.Update(&err.Transaction)
 		}
 		return roachpb.NewError(err)
 	}
@@ -1069,8 +1066,7 @@ func (tc *TxnCoordSender) updateStateLocked(
 			PrevError: pErr.String(),
 		})
 		// Cleanup.
-		cp := errTxn.Clone()
-		tc.mu.txn.Update(&cp)
+		tc.mu.txn.Update(errTxn)
 		tc.cleanupTxnLocked(ctx)
 	}
 	return pErr

--- a/pkg/roachpb/api.go
+++ b/pkg/roachpb/api.go
@@ -383,10 +383,13 @@ func (h *BatchResponse_Header) combine(o BatchResponse_Header) error {
 		)
 	}
 	h.Timestamp.Forward(o.Timestamp)
-	if h.Txn == nil {
-		h.Txn = o.Txn
-	} else {
-		h.Txn.Update(o.Txn)
+	if txn := o.Txn; txn != nil {
+		if h.Txn == nil {
+			txnClone := txn.Clone()
+			h.Txn = &txnClone
+		} else {
+			h.Txn.Update(txn)
+		}
 	}
 	h.Now.Forward(o.Now)
 	h.CollectedSpans = append(h.CollectedSpans, o.CollectedSpans...)

--- a/pkg/roachpb/batch.go
+++ b/pkg/roachpb/batch.go
@@ -424,7 +424,6 @@ func (br *BatchResponse) Combine(otherBatch *BatchResponse, positions []int) err
 			return errors.Errorf("can not combine %T and %T", valLeft, valRight)
 		}
 	}
-	br.Txn.Update(otherBatch.Txn)
 	return nil
 }
 

--- a/pkg/roachpb/batch.go
+++ b/pkg/roachpb/batch.go
@@ -72,17 +72,17 @@ func (ba *BatchRequest) SetActiveTimestamp(nowFn func() hlc.Timestamp) error {
 // UpdateTxn updates the batch transaction from the supplied one in
 // a copy-on-write fashion, i.e. without mutating an existing
 // Transaction struct.
-func (ba *BatchRequest) UpdateTxn(otherTxn *Transaction) {
-	if otherTxn == nil {
+func (ba *BatchRequest) UpdateTxn(o *Transaction) {
+	if o == nil {
 		return
 	}
-	otherTxn.AssertInitialized(context.TODO())
+	o.AssertInitialized(context.TODO())
 	if ba.Txn == nil {
-		ba.Txn = otherTxn
+		ba.Txn = o
 		return
 	}
 	clonedTxn := ba.Txn.Clone()
-	clonedTxn.Update(otherTxn)
+	clonedTxn.Update(o)
 	ba.Txn = &clonedTxn
 }
 

--- a/pkg/roachpb/errors.go
+++ b/pkg/roachpb/errors.go
@@ -171,9 +171,9 @@ func (e *Error) GetTxn() *Transaction {
 	return e.UnexposedTxn
 }
 
-// UpdateTxn updates the txn.
+// UpdateTxn updates the error transaction.
 func (e *Error) UpdateTxn(o *Transaction) {
-	if e == nil {
+	if o == nil {
 		return
 	}
 	if e.UnexposedTxn == nil {

--- a/pkg/roachpb/errors.go
+++ b/pkg/roachpb/errors.go
@@ -151,7 +151,6 @@ func (e *Error) GoError() error {
 
 // SetTxn sets the txn and resets the error message. txn is cloned before being
 // stored in the Error.
-// TODO(kaneda): Unexpose this method and make callers use NewErrorWithTxn.
 func (e *Error) SetTxn(txn *Transaction) {
 	e.UnexposedTxn = txn
 	if txn != nil {

--- a/pkg/storage/store.go
+++ b/pkg/storage/store.go
@@ -2795,18 +2795,13 @@ func (s *Store) Send(
 			// this node.
 			if pErr != nil {
 				pErr.OriginNode = ba.Replica.NodeID
-				if txn := pErr.GetTxn(); txn != nil {
-					// Clone the txn, as we'll modify it.
-					pErr.SetTxn(txn)
-				} else {
+				if txn := pErr.GetTxn(); txn == nil {
 					pErr.SetTxn(ba.Txn)
 				}
-				pErr.GetTxn().UpdateObservedTimestamp(ba.Replica.NodeID, now)
 			} else {
 				if br.Txn == nil {
 					br.Txn = ba.Txn
 				}
-				br.Txn.UpdateObservedTimestamp(ba.Replica.NodeID, now)
 				// Update our clock with the outgoing response txn timestamp
 				// (if timestamp has been forwarded).
 				if ba.Timestamp.Less(br.Txn.Timestamp) {

--- a/pkg/testutils/storageutils/mocking.go
+++ b/pkg/testutils/storageutils/mocking.go
@@ -58,11 +58,7 @@ func WrapFilterForReplayProtection(
 func shallowCloneErrorWithTxn(pErr *roachpb.Error) *roachpb.Error {
 	if pErr != nil {
 		pErrCopy := *pErr
-		txn := pErrCopy.GetTxn()
-		if txn != nil {
-			txnClone := txn.Clone()
-			pErrCopy.SetTxn(&txnClone)
-		}
+		pErrCopy.SetTxn(pErrCopy.GetTxn())
 		return &pErrCopy
 	}
 

